newIDE/app/src/InstancesEditor/index.js

```diff
@@ -804,16 +811,69 @@ export default class InstancesEditor extends Component<Props> {
     this.viewPosition.scrollTo(x, y);
   }
 
-  centerView() {
+  zoomToFitContent() {
+    const instanceMeasurer = this.instancesRenderer.getInstanceMeasurer();
+    let rectangle: ?Rectangle;
+    const getInstanceRectangle = new gd.InitialInstanceJSFunctor();
+    // $FlowFixMe - invoke is not writable
+    getInstanceRectangle.invoke = instancePtr => {
+      // $FlowFixMe - wrapPointer is not exposed
+      const instance = gd.wrapPointer(instancePtr, gd.InitialInstance);
+      if (!rectangle) {
+        rectangle = instanceMeasurer.getInstanceAABB(instance, new Rectangle());
+      } else {
+        rectangle.union(
+          instanceMeasurer.getInstanceAABB(instance, new Rectangle())
+        );
+      }
+    };
+    const { initialInstances } = this.props;
+    if (initialInstances.getInstancesCount() > 0) {
+      // $FlowFixMe - JSFunctor is incompatible with Functor
+      initialInstances.iterateOverInstances(getInstanceRectangle);
+
+      if (rectangle) {
+        const idealZoom = this.viewPosition.fitToRectangle(rectangle);
+        this.setZoomFactor(idealZoom);
+      }
+    }
+  }
+
+  zoomToInitialPosition() {
     const x = this.props.project.getGameResolutionWidth() / 2;
     const y = this.props.project.getGameResolutionHeight() / 2;
     this.viewPosition.scrollTo(x, y);
+    this.setZoomFactor(1);
+  }
+
+  zoomToFitSelection(instances: Array<gdInitialInstance>) {
+    if (instances.length === 0) return;
+    const [firstInstance, ...otherInstances] = instances;
+    const instanceMeasurer = this.instancesRenderer.getInstanceMeasurer();
+    let selectedInstancesRectangle = instanceMeasurer.getInstanceAABB(
+      firstInstance,
+      new Rectangle()
+    );
+    otherInstances.forEach(instance => {
+      selectedInstancesRectangle.union(
+        instanceMeasurer.getInstanceAABB(instance, new Rectangle())
+      );
+    });
+    const idealZoom = this.viewPosition.fitToRectangle(
+      selectedInstancesRectangle
+    );
+    this.setZoomFactor(idealZoom);
   }
 
-  centerViewOn(instances: Array<gdInitialInstance>) {
-    if (!instances.length) return;
+  centerViewOnLastInstance(instances: Array<gdInitialInstance>) {
+    if (instances.length === 0) return;
 
-    this.viewPosition.scrollToInstance(instances[instances.length - 1]);
+    const instanceMeasurer = this.instancesRenderer.getInstanceMeasurer();
+    let lastInstanceRectangle = instanceMeasurer.getInstanceAABB(
+      instances[instances.length - 1],
+      new Rectangle()
+    );
+    this.viewPosition.fitToRectangle(lastInstanceRectangle);
```

Comment:
Does it fit to the whole selection or just one instance of the selection?

Reply:
I kept the current behaviour: the camera centers on the last selected instance. It's kinda logic since you're selecting them from the instance list, it shows on the canvas what is the instance that you just selected

Reply:
I'm not sure it's a good UX. The view will jump from one instance to another and other instances from the selection will go out of the view. User won't have a clear view of what is selected.

Reply:
I would say:

- Camera zooms in to selected object on the scene
- It does change the zoom (it zooms in to the selected object on the scene).

As for instances repeated across the scene: I see the issue.
Indeed at first I would say "zoom to the instance selected on the object list", but considering that there could be a "duplicated" instance on the scene, I would go to what design software does:
It doesn't allow you to "Zoom to instance with is selected from the object list", but rather selecting an object from the list and "Highlight instances of this object on the project" on the same view, without zooming-in or out.
The copy that they use on the "right-click menu" is "Highlight component instances" or for GDevelop: "Highlight object instances".